src/Nethermind/Nethermind.Monitoring/Metrics/MetricsController.cs

```diff
@@ -119,56 +215,108 @@ private static string GetStaticMemberInfo(Type givenInformer, string givenName)
             object value = info.GetValue(null) ?? throw new NotSupportedException("Developer error: a requested static description field was not initialised!");
-            return value.ToString();
+            return value.ToString()!;
         }
 
         private void EnsurePropertiesCached(Type type)
         {
-            static bool NotEnumerable(Type t) => !t.IsAssignableTo(typeof(System.Collections.IEnumerable));
-
-            static Func<double> GetValueAccessor(MemberInfo member)
+            if (!_metricUpdaters.ContainsKey(type))
             {
-                if (member is PropertyInfo property)
+                Meter? meter = null;
+                if (_useCounters)
                 {
-                    return () => Convert.ToDouble(property.GetValue(null));
+                    meter = new(type.Namespace!);
                 }
 
-                if (member is FieldInfo field)
+                IList<IMetricUpdater> metricUpdaters = new List<IMetricUpdater>();
+                foreach (var propertyInfo in type.GetProperties())
+                {
+                    if (TryCreateMetricUpdater(type, meter, propertyInfo, out IMetricUpdater updater))
+                    {
+                        metricUpdaters.Add(updater);
+                    }
+                }
+                foreach (var fieldInfo in type.GetFields())
                 {
-                    return () => Convert.ToDouble(field.GetValue(null));
+                    if (TryCreateMetricUpdater(type, meter, fieldInfo, out IMetricUpdater updater))
+                    {
+                        metricUpdaters.Add(updater);
+                    }
                 }
 
-                throw new NotImplementedException($"Type of {member} is not handled");
+                _metricUpdaters[type] = metricUpdaters.ToArray();
             }
+        }
 
-            if (!_membersCache.ContainsKey(type))
+        private bool TryCreateMetricUpdater(Type type, Meter? meter, MemberInfo memberInfo, out IMetricUpdater metricUpdater)
+        {
+            Type memberType;
+            if (memberInfo is PropertyInfo property)
+            {
+                memberType = property.PropertyType;
+            }
+            else if (memberInfo is FieldInfo field)
             {
-                _membersCache[type] = type.GetProperties()
-                    .Where(p => NotEnumerable(p.PropertyType))
-                    .Concat<MemberInfo>(type.GetFields().Where(f => NotEnumerable(f.FieldType)))
-                    .Select(member => (member, GetGaugeNameKey(type.Name, member.Name), GetValueAccessor(member)))
-                    .ToArray();
+                memberType = field.FieldType;
+            }
+            else
+            {
+                throw new UnreachableException();
```

Comment:
Extract to method (potentially in some core utils extension class? `MemberInfoExtensions`?